pkg/apis/internal.acorn.io/v1/map.go

```diff
@@ -3,25 +3,98 @@ package v1
 import (
 	"bytes"
 	"encoding/json"
+	"fmt"
 
+	"github.com/acorn-io/baaah/pkg/typed"
 	"github.com/sirupsen/logrus"
+	"k8s.io/kube-openapi/pkg/common"
+	"k8s.io/kube-openapi/pkg/validation/spec"
 )
 
-type GenericMap map[string]interface{}
+type GenericMap struct {
+	// +optional
+	Data map[string]any `json:"-"`
+}
+
+func (_ GenericMap) OpenAPIDefinition() common.OpenAPIDefinition {
+	return common.OpenAPIDefinition{
+		Schema: spec.Schema{
+			VendorExtensible: spec.VendorExtensible{
+				Extensions: spec.Extensions{
+					"x-kubernetes-preserve-unknown-fields": "true",
+				},
+			},
+			SchemaProps: spec.SchemaProps{
+				Type: []string{"object"},
+			},
+		},
+	}
+}
+
+func (g *GenericMap) UnmarshalJSON(data []byte) error {
+	if g == nil {
+		return fmt.Errorf("%T: UnmarshalJSON on nil pointer", g)
+	}
+
+	dec := json.NewDecoder(bytes.NewBuffer(data))
+	dec.UseNumber()
+
+	d := map[string]any{}
+	if err := dec.Decode(&d); err != nil {
+		return fmt.Errorf("%T: Failed to decode data: %w", g, err)
+	}
+
+	if _, err := translateObject(&d); err != nil {
+		return fmt.Errorf("%T: Failed to translate object: %w", g, err)
+	}
+
+	if len(d) > 0 {
+		// Consumers expect empty generic maps to have a nil Data field
+		g.Data = d
+	}
+
+	return nil
+}
+
+// GetData returns the underlying map[string]any and nil if the GenericMap is nil.
+func (g *GenericMap) GetData() map[string]any {
+	if g == nil {
+		return nil
+	}
+
+	return g.Data
+}
+
+// Merge merges the given map into this map, returning a new map, leaving the original unchanged.
+func (g *GenericMap) Merge(from *GenericMap) *GenericMap {
+	if from == nil || from.Data == nil {
+		return g
+	}
+
+	if g == nil {
+		g = &GenericMap{}
+	}
+
+	return &GenericMap{
+		Data: typed.Concat(g.Data, from.Data),
+	}
+}
 
-func (in GenericMap) MarshalJSON() ([]byte, error) {
-	return json.Marshal((map[string]interface{})(in))
+// MarshalJSON may get called on pointers or values, so implement MarshalJSON on value.
+// http://stackoverflow.com/questions/21390979/custom-marshaljson-never-gets-called-in-go
```

Comment:
nit: It is enough to leave your comment. The stack overflow link doesn't provide enough value here.